kedro/framework/project/__init__.py

```diff
@@ -39,6 +41,25 @@ def validate(self, settings, *args, **kwargs):
                 )
 
 
+class _HasSharedParentClassValidator(Validator):
+    """A validator to check that the parent of the default class is an ancestor of
+    the settings value."""
+
+    def validate(self, settings, *args, **kwargs):
+        super().validate(settings, *args, **kwargs)
+
+        default_class = self.default(settings, self)
+        for name in self.names:
+            setting_value = getattr(settings, name)
+            default_class_parent = default_class.mro()[1]
+            if default_class_parent not in setting_value.mro():
+                raise ValidationError(
+                    f"Invalid value `{setting_value.__module__}.{setting_value.__qualname__}` "
+                    f"received for setting `{name}`. It must be a subclass of "
+                    f"`{default_class_parent.__module__}.{default_class_parent.__qualname__}`."
```

Comment:
Maybe we could add some comments to explain this.

Here is my understanding, `default_class` is `ConfigLoader` if we are talking about config_loader, but this is expected to be a generic validator.

So the assumption here is, **the default_class is something inherit the `AbstractXXXXX` class, thus default_class.mro()[1] will be pointing to the `AbstractXXXXX` class.** If the default aren't directly inherit from the abstract class, this would fail.

Reply:
Yes, you're exactly right and I will add a comment to explain this!

`ConfigLoader.mro()` looks like this:
```
[kedro.config.config.ConfigLoader,
 kedro.config.abstract_config.AbstractConfigLoader,
 abc.ABC,
 object]
```
So we are picking out the direct parent of `ConfigLoader`,  whereas we look at all ancestors of the class specified in settings.py. This means that if I make a `CustomConfigLoader` that inherits from `TemplatedConfigLoader`, it will still pass the validation.

I actually originally wrote this to check for a common ancestor anywhere in the inheritance chain, like `set(default_class.mro()) & set(setting_value.mro()) - {abc.ABC, object}` but decided that was more confusing overall because you have to explicitly exclude some classes. Like you say, I think the assumption that `default_class`'s direct parent is `AbstractXXX` is unlikely to be false in the future. And if we do change it then the tests will fail straight away and we can change the validation code then.

Reply:
Yes, I agree just looking 1 layer up is sufficient.